OSDOCS-2415: Adding mTLS info (#35473)
openshift/openshift-docs: modules/nw-mutual-tls-auth.adoc

// Module included in the following assemblies:
//
// * ingress/ingress-operator.adoc

[id=nw-mutual-tls-auth_{context}]
= Configuring Mutual TLS Authentication

> You can configure the Ingress Controller to enable mutual clientTLS(mTLS) authentication by setting `spec.clientTLS`. The clientTLS configures the Ingress controller to verify client certificates (CAs), certificate revocation list (CRL), and/or any certificate subject filters. Requests that do not provide valid certificates are rejected.

Reviewer: The description might be clearer if it avoided mixing "client TLS" and "mutual TLS".  The Route API allows configuring server TLS, this new IngressController API allows configuring client TLS, and client and server TLS together constitute mutual TLS.  

The user can use the API to configure a CA certificate that is used to sign client certificates.  The operator downloads any associated CRL (the user doesn't specify the CRL directly).  
```suggestion
You can configure the Ingress Controller to enable mutual TLS (mTLS) authentication by setting `spec.clientTLS`. The `clientTLS` field specifies configuration for the Ingress Controller to verify client certificates.  This configuration includes the CA certificate that is used to sign client certificates, as well as an optional list of certificate subject filters. Requests that do not provide valid certificates are rejected. If the CA certificate specifies an X509v3 certificate revocation list (CRL) distribution point, the Ingress Controller downloads and enforces the CRL.
```

.Prerequisites

* You have access to the cluster as a user with the cluster-admin role.
* A ConfigMap that is in the openshift-config namespace.

.Procedure

Configure Mutual TLS authentication using the `spec.ClientTLS` field for the Ingress Controller.

.  Edit the `IngressController` resource in the `openshift-ingress-operator` project:
[source,terminal]
----
$ oc edit IngressController default -n openshift-ingress-operator
----

.  Edit the yaml:

.Sample `IngressController` CR for a `clientTLS` profile that specifies filtering patterns.

[source,yaml]
----
  apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: default
    namespace: openshift-ingress-operator
  spec:
    clientTLS:
      clientCertificatePolicy: Required
      clientCA:

>         name: router-ca-certs-default

Reviewer: There needs to be a step for creating the configmap specified here.  For example, the following command should work:
```
oc create configmap router-ca-certs-default --from-file=ca-bundle.pem=client-ca.crt -n openshift-config
```
It is important to note that the configmap data key needs to be `ca-bundle.pem`, and the data value needs to be a CA certificate in PEM format.

// Module included in the following assemblies:
//
// * ingress/ingress-operator.adoc

[id=nw-mutual-tls-auth_{context}]
= Configuring Mutual TLS Authentication

You can configure the Ingress Controller to enable mutual TLS (mTLS) authentication by setting `spec.clientTLS`. The `clientTLS` field specifies configuration for the Ingress Controller to verify client certificates. This configuration includes the CA certificate, which is used to sign client certificates, as well as an optional list of certificate subject filters. If the CA certificate specifies an X509v3 certificate revocation list (CRL) distribution point, the Ingress Controller downloads and enforces the CRL. Requests that do not provide valid certificates are rejected.

.Prerequisites

* You have access to the cluster as a user with the cluster-admin role.

.Procedure

Configure Mutual TLS authentication using the `spec.ClientTLS` field for the Ingress Controller.

> . Create a ConfigMap that is in the openshift-config namespace:

Reviewer 2: Add an extra line between Line 17 and Line 18. The Preview is off.

+
[source,terminal]
----

> oc create configmap router-ca-certs-default --from-file=ca-bundle.pem=client-ca.crt -n openshift-config

Reviewer 2: ```suggestion
$ oc create configmap router-ca-certs-default --from-file=ca-bundle.pem=client-ca.crt -n openshift-config
```

----
+
[NOTE]
====
Configmap data key must follow a naming convention, such as `ca-bundle.pem`, and the data value must be a CA certificate in PEM format.
====
.  Edit the `IngressController` resource in the `openshift-ingress-operator` project:
+
[source,terminal]
----
$ oc edit IngressController default -n openshift-ingress-operator
----

.  Edit the yaml:

> .Sample `IngressController` CR for a `clientTLS` profile that specifies filtering patterns.

Reviewer 2: ```suggestion
.Sample `IngressController` CR for a `clientTLS` profile that specifies filtering patterns
```

Removing the period at the end of your sample title.

// Module included in the following assemblies:
//
// * ingress/ingress-operator.adoc

[id=nw-mutual-tls-auth_{context}]
= Configuring Mutual TLS Authentication

> You can configure the Ingress Controller to enable mutual TLS (mTLS) authentication by setting a `spec.clientTLS` value. The `clientTLS` field specifies configuration for the Ingress Controller to verify client certificates. This configuration includes setting a `clientCA` value, which is a reference to a config map that contains the PEM-encoded CA certificate bundle that is used to verify a client's certificate, as well as an optional list of certificate subject filters.

Author: I'll take all these, yep.

> If the `clientCA` value specifies an X509v3 certificate revocation list (CRL) distribution point, the Ingress Controller downloads and enforces the CRL. Requests that do not provide valid certificates are rejected.

Reviewer: Technically, it's the operator that downloads the CRL, not the ingress controller.  
```suggestion
If the `clientCA` value specifies an X509v3 certificate revocation list (CRL) distribution point, the Ingress Operator downloads the CRL and configures the Ingress Controller to respect it. Requests that do not provide valid certificates are rejected.
```

* You have access to the cluster as a user with the cluster-admin role.

.Procedure

Configure Mutual TLS authentication using the `spec.ClientTLS` field for the Ingress Controller.

> . Create a ConfigMap that is in the openshift-config namespace:

Reviewer 3: I can't tell if you overlooked Kathryn's advice or believe there is an issue.  Please revise as Kathryn suggested:

". Crate a config map in the `openshift-config` namespace:"

+
[source,terminal]
----
$ oc create configmap router-ca-certs-default --from-file=ca-bundle.pem=client-ca.crt -n openshift-config
----
+
[NOTE]
====

> The config map data key must follow a naming convention, such as `ca-bundle.pem`, and the data value must be a CA certificate in PEM format.

Reviewer: It isn't exactly a convention, but rather a specific key that must be used. 
```suggestion
The config map data key must be `ca-bundle.pem`, and the data value must be a CA certificate in PEM format.
```

Reviewer 3: Miciah's suggestion is a good one.

====
.  Edit the `IngressController` resource in the `openshift-ingress-operator` project:
+
[source,terminal]
----
$ oc edit IngressController default -n openshift-ingress-operator
----

> .  Edit the yaml:

Reviewer 4: ```suggestion
. Edit the yaml:
```

I think you could be clearer what you're adding to the resource.

.Sample `IngressController` CR for a `clientTLS` profile that specifies filtering patterns

> 

Reviewer 3: nit: The convention seems to be the block title, no blank line, and then the `[source,...]` line, like the first `.Example output` in the [Code blocks, command syntax, and example output](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) in the doc guidelines.

// Module included in the following assemblies:
//
// * ingress/ingress-operator.adoc

[id=nw-mutual-tls-auth_{context}]

> = Configuring Mutual TLS Authentication

Reviewer 3: Apologies for missing this earlier, please use sentence case in [headings](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings).

You can configure the Ingress Controller to enable mutual TLS (mTLS) authentication by setting a `spec.clientTLS` value. The `clientTLS` value configures the Ingress Controller to verify client certificates. This configuration includes setting a `clientCA` value, which is a reference to a config map. The config map contains the PEM-encoded CA certificate bundle that is used to verify a client's certificate. Optionally, you can configure a list of certificate subject filters.

If the `clientCA` value specifies an X509v3 certificate revocation list (CRL) distribution point, the Ingress Operator downloads the CRL and configures the Ingress Controller to acknowledge it. Requests that do not provide valid certificates are rejected.

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.

.Procedure

> Configure Mutual TLS authentication using the `spec.ClientTLS` value for the Ingress Controller.

Reviewer 3: Regarding line 17, first, it is essentially a repeat of the perfectly clear first sentence on line 8, so it doesn't seem necessary.  Secondly, the mod docs template for a [procedure](https://raw.githubusercontent.com/redhat-documentation/modular-docs/master/modular-docs-manual/files/TEMPLATE_PROCEDURE_doing-one-procedure.adoc) doesn't have a place for an intro one-liner after the `.Prerequisites` or `.Procedure` block title.

Removing it is fine.  Your procedure is perfectly clear.

> . Create a ConfigMap that is in the `openshift-config` namespace:

Reviewer 3: Please replace ConfigMap with "config map" because it is a [general reference](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-general-references).

+
[source,terminal]
----
$ oc create configmap router-ca-certs-default --from-file=ca-bundle.pem=client-ca.crt -n openshift-config
----
+
[NOTE]
====
The config map data key must be `ca-bundle.pem`, and the data value must be a CA certificate in PEM format.
====

. Edit the `IngressController` resource in the `openshift-ingress-operator` project:
+
[source,terminal]
----
$ oc edit IngressController default -n openshift-ingress-operator
----

> . Edit the yaml:

Reviewer 3: Maybe ". Add the `spec.clientTLS` field and subfields to configure mutual TLS:"

If you prefer to keep what you have, consider replacing "yaml" with either "the manifest" or "YAML".

+
.Sample `IngressController` CR for a `clientTLS` profile that specifies filtering patterns

> +

Reviewer 3: Line 41 is unnecessary.  If you remove it, the file will match the conventions from the [code blocks](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) section of the docs guidelines.  See how the `.Example output` block title immediate precedes `[source,...]`.

// Module included in the following assemblies:
//
// * ingress/ingress-operator.adoc

[id=nw-mutual-tls-auth_{context}]
= Configuring Mutual TLS Authentication

You can configure the Ingress Controller to enable mutual TLS(mTLS) authentication. mTLS verifies the client certificates (CAs), the certificate revocation list (CRL), and/or any certificate subject filters and rejects requests from clients that do not provide valid certificates.

The {product-title} Ingress Controller configures HAProxy using a configuration template.  The

> template uses environment variables as input parameters. You can configure mTLS using the following environment variables:

Reviewer 5: User is just allowed to set `spec.clientTLS` in ingresscontroller to configure mTLS, the ingress operator will watch the `spec.clientTLS` and update those environment variables accordingly. 
Please help check with Devel if the description of these environment variables are required in the document. Thanks.

Author: I suspect you're correct. I wondered if this was behind the scenes myself but included it anyways. @Miciah can say for sure.

Reviewer: Hongan is right, the environment variables are an implementation detail that isn't necessary to document here.
